### ISSUE TITLE
Test fails with `unknown_failure decoding unicode is not supported`

### ISSUE THREAD

**Issue by hellais** [ooni/probe#751]
### Description

I noticed that recently the whatsapp test is failing with the following error:

```
2017-04-28 01:34:53,812411+0000 [HTTP11ClientProtocol,client] [!] Failed to connect to http://149.154.171.5:443: unknown_failure decoding Unicode is not supported
```

As result this is what the result page for the whatsapp test looks like:
![screen shot 2017-04-28 at 12 13 00 0000](https://cloud.githubusercontent.com/assets/424620/25528254/5f4742c8-2c0c-11e7-9413-a0d450f6e7af.png)

### Further information

**How did you install ooniprobe**

This is running on lepidopter 1.0.0

**What operating system are you running and what version is it?**

<details><summary>Output of `ooniprobe --version`</summary>

```
ooniprobe version: 2.2.0
```
</details>

**Comment by hellais** [comment]
Strangely it appears that the same version of ooniprobe (2.2.0) on macOS does not exhibit the same issue when running the whatsapp test from the command line.

**Comment by anadahz** [comment]
I was able to reproduce this in lepidopter, (with ooniprobe 2.2.0):
https://explorer.ooni.torproject.org/measurement/20170427T005245Z_AS1916_Kgh3TX6DsOxHjHwtEusQylNNKogzKySGqgjbH56Aacmqnrgtmu

**Comment by hellais** [comment]
So I have come to the bottom of this issue. The root cause of it is txtorcon 0.19.0 (see: https://github.com/meejah/txtorcon/issues/232).

This issue would have however been mitigated if the code in question was a bit more defensive and didn't make too many assumptions on the types of values.

Here is an explanation of what is going on exactly:

The ooni function that is failing is `representBody` (https://github.com/TheTorProject/ooni-probe/blob/master/ooni/common/http_utils.py#L5) that is making the assumption that the argument it gets passed is going to be a str, but in our case it happens to instead be a unicode.

This function is actually applied also to response headers and not only to bodies.

The reason why it is a unicode (HTTP bodies, nor headers, should never be unicode actually) is that response headers are stripped from the probes IP address [here](https://github.com/TheTorProject/ooni-probe/blob/master/ooni/templates/httpt.py#L178) and [here](https://github.com/TheTorProject/ooni-probe/blob/master/ooni/templates/httpt.py#L181). Normally this should be fine, because we are just calling replace on a string with some strings, however it seems like python will cast to unicode if one of the two arguments is unicode (even if not replacement occurs!).

See this as an example:

```
>>> a = "antani"
>>> repr(a)
"'antani'"
>>> a = a.replace("notexist", "bar")
>>> repr(a)
"'antani'"
>>> a = a.replace(u"notexist", "bar")
>>> repr(a)
"u'antani'"
```

The address value of probe_ip will in fact be of unicode type (and we don't enforce str type anywhere) ONLY when the lookup is done via tor (because ooniprobe started tor) and txtorcon is >= 0.19.0, because of the forementioned bug.

At this point I think that the quick workaround to this issues is probably that of setting the txtorcon version in lepidopter to 0.18.0 that is known to work without any issues.

We should moreover also do fixes inside of ooniprobe to:

* [ ] Enforce that the probe_ip is always str and should never be unicode

* [ ] Make `representBody` more robust to handling unicode as input and if such input is detected it should just not encode it again.

**Comment by anadahz** [comment]
> At this point I think that the quick workaround to this issues is probably that of setting the txtorcon version in lepidopter to 0.18.0 that is known to work without any issues.

I was able to reproduce https://github.com/meejah/txtorcon/issues/232 in a non lepidopter platform (x86/64) and I guess this bug could be possibly reproduced in MacOS. That said and in order to avoid all possible surprises.I think it makes sense to pin the version of txtorcon to 0.18.0 in ooniprobe (`requirements`.txt file) rather than only to lepidopter.

**Comment by hellais** [comment]
> I was able to reproduce meejah/txtorcon#232 in a non lepidopter platform (x86/64) and I guess this bug could be possibly reproduced in MacOS. 

On macOS homebrew the version of txtorcon is set to a working one by downloading the dependency at a correct version.

> That said and in order to avoid all possible surprises.I think it makes sense to pin the version of txtorcon to 0.18.0 in ooniprobe (requirements.txt file) rather than only to lepidopter.

That would require making a new release of ooniprobe that at that point makes most sense to make with the fix in #752. Also pinning versions in requirements.txt is tricky since we need to ensure the allowed versions include the version in the various versions of debian we wish to support. I would say for the time being we proceed with pinning the version in lepidopter and address things on the ooniprobe side with a new release once #752 is merged and a new release is cut (I would like the new release to also include the new translations)

**Comment by anadahz** [comment]
Currently there are a number of lepidopter users affected so it makes sense to make a [release](https://github.com/TheTorProject/lepidopter/pull/97) with txtorcon `0.18.0` and publish the lepidopter [update](https://github.com/OpenObservatory/lepidopter-update/pull/22).

> address things on the ooniprobe side with a new release once #752 is merged and a new release is cut (I would like the new release to also include the new translations)

I will aim for the #752 to be merged and release early as from what I see in random reports this bug potentially affects/will affect Linux platforms even if the new translation are not ready yet.

**Comment by hellais** [comment]
> I will aim for the #752 to be merged and release early as from what I see in random reports this bug potentially affects/will affect Linux platforms even if the new translation are not ready yet.

It will only affect people that install it with pip, but the version of txtorcon in debian is a safe one, so it should not affect them.

**Comment by anadahz** [comment]
This bug applies also to the telegram test, makes sense since it's almost the same test --an [example report](https://s3.amazonaws.com/ooni-public/sanitised/2017-07-19/20170718T045628Z-ID-AS23700-telegram-20170718T045631Z_AS23700_kCH2VLliOLF7Y2sh6cEIM0AtAuFnW0jRtctLFTauTD0DfnEdxj-0.2.0-probe.json).
Shall I open a new issue for it or are we tracking it here?

**Comment by hellais** [comment]
> This bug applies also to the telegram test, makes sense since it's almost the same test --an example report.

It probably applies to any test. The title is misleading, because it implies it's somehow whatsapp specific. It isn't.

> Shall I open a new issue for it or are we tracking it here?

What would be useful is testing if the patch I proposed fixes the bug in all cases.

**Comment by hellais** [comment]
Issue moved to [ooni/probe-legacy #38](https://github.com/ooni/probe-legacy/issues/38) via [**ZenHub**](https://www.zenhub.com/)
